File: mobile-app/packages/services/test/news_service_parity_test.dart
import 'dart:convert';
import 'package:test/test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:smwa_services/services.dart';

void main() {
  test('ledger increments again after window expiry', () async {
    var calls = 0;
    final ledger = ApiQuotaLedger(1, const Duration(milliseconds: 50));
    final client = MockClient((req) async {
      calls++;
      return http.Response(
          jsonEncode({
            'results': [
              {'title': 't', 'link': 'l', 'source_id': 's', 'pubDate': 'p'}
            ]
          }),
          200);
    });
    final svc = NewsService('k', ledger: ledger, client: client);
    await svc.getDigest('aa');
    await Future.delayed(const Duration(milliseconds: 60));
    await svc.getDigest('bb');
    expect(calls, 2);
  });

  test('skips request when ledger disallows', () async {
    var calls = 0;
    final ledger = ApiQuotaLedger(0);
    final client = MockClient((req) async {
      calls++;
      return http.Response('', 200);
    });
    final svc = NewsService('k', ledger: ledger, client: client);
    final res = await svc.getDigest('aa');
    expect(res, isNull);
    expect(calls, 0);
  });

  test('ledger increments on successful request', () async {
    final ledger = ApiQuotaLedger(1);
    final client = MockClient((req) async => http.Response(
        jsonEncode({
          'results': [
            {'title': 't', 'link': 'l', 'source_id': 's', 'pubDate': 'p'}
          ]
        }),
        200));
    final svc = NewsService('k', ledger: ledger, client: client);
    await svc.getDigest('aa');
    expect(ledger.isSafe(), isFalse);
  });

  test('caches result for 12h', () async {
    var calls = 0;
    final ledger = ApiQuotaLedger(2);
    final client = MockClient((req) async {
      calls++;
      return http.Response(
          jsonEncode({
            'results': [
              {'title': 't', 'link': 'l', 'source_id': 's', 'pubDate': 'p'}
            ]
          }),
          200);
    });
    final svc = NewsService('k', ledger: ledger, client: client);
    await svc.getDigest('aa');
    await svc.getDigest('aa');
    expect(calls, 1);
  });

  test('falls back to rss when api fails', () async {
    var calls = 0;
    final rss =
        '<?xml version="1.0"?><rss><channel><item><title>r</title><link>u</link><pubDate>d</pubDate></item></channel></rss>';
    final client = MockClient((req) async {
      calls++;
      if (calls == 1) return http.Response('', 500);
      return http.Response(rss, 200);
    });
    final svc = NewsService('k', ledger: ApiQuotaLedger(1), client: client);
    final news = await svc.getDigest('aa');
    expect(news?.first['title'], 'r');
    expect(calls, 2);
  });
}

File: web-app/tests/NewsServiceParity.test.ts
import { describe, it, expect, vi } from 'vitest';
import { NewsService } from '../src/services/NewsService';
import { NetClient, HALF_DAY_MS } from '../../packages/core/net';

function setupService() {
  const svc = new NewsService('k');
  const ledger = { isSafe: vi.fn().mockReturnValue(true), increment: vi.fn() };
  (svc as any).ledger = ledger;
  return { svc, ledger };
}

describe('NewsService parity with Dart', () => {
  it('increments ledger on successful request', async () => {
    const { svc, ledger } = setupService();
    const client = {
      get: vi.fn().mockImplementation(async () => {
        ledger.increment();
        return [{ title: 't', url: 'l', source: 's', published: 'p' }];
      })
    } as unknown as NetClient;
    (svc as any).client = client;

    const news = await svc.getNews('aa');
    expect(news?.[0].title).toBe('t');
    expect(client.get).toHaveBeenCalled();
    expect(ledger.increment).toHaveBeenCalledTimes(1);
  });

  it('passes HALF_DAY_MS ttl to NetClient', async () => {
    const { svc } = setupService();
    const client = { get: vi.fn().mockResolvedValue([]) } as unknown as NetClient;
    (svc as any).client = client;
    await svc.getNews('bb');
    expect(client.get).toHaveBeenCalledWith(
      expect.stringContaining('bb'),
      (svc as any).cache,
      expect.any(Function),
      HALF_DAY_MS
    );
  });

  it('falls back to RSS when API fails', async () => {
    const { svc, ledger } = setupService();
    const rss = '<?xml version="1.0"?><rss><channel><item><title>r</title><link>u</link><pubDate>d</pubDate></item></channel></rss>';
    const fetchMock = vi
      .fn()
      .mockResolvedValueOnce({ ok: false })
      .mockResolvedValueOnce({ ok: true, text: async () => rss });
    global.fetch = fetchMock as any;

    const news = await svc.getNews('cc');
    expect(news).toEqual([{ title: 'r', url: 'u', source: 'rss', published: 'd' }]);
    expect(fetchMock).toHaveBeenCalledTimes(2);
    expect(ledger.increment).not.toHaveBeenCalled();
  });
});
